### ISSUE TITLE
When UnzipUtil.createZipInputStream(ZipModel, FileHeader, char[]) throws an Exception it doesn't close InputStreams

### ISSUE THREAD

**Issue by daelfdude** [srikanth-lingala/zip4j#58]
I have a unit test asserting that a file is password protected.  As an assertion on that, I check that opening the zip file with a null password fails.  However when I do that, the file being asserted upon becomes undeletable due to the unclosed input streams this method creates.

**Comment by srikanth-lingala** [comment]
I used the following code to test, and this code deletes the zip file. I tested it on a mac though. Which os are you using? Does your tests (or user running the tests) have the rights to delete a file?

```
ZipFile zipFile = new ZipFile("/Users/slingala/Downloads/test.zip", "password".toCharArray());
    ZipParameters zipParameters = new ZipParameters();
    zipParameters.setEncryptFiles(true);
    zipParameters.setEncryptionMethod(EncryptionMethod.AES);
    zipFile.addFile(TestUtils.getTestFileFromResources("sample.pdf"), zipParameters);

    try {
      zipFile = new ZipFile("/Users/slingala/Downloads/test.zip");
      zipFile.extractAll("/Users/slingala/Downloads/");
    } catch (Exception e) {
      assertThat(zipFile.getFile().delete()).isTrue();
      throw e;
    }
```

**Comment by daelfdude** [comment]
I’m using Windows.  And if I remove the call to 
UnzipUtil.createZipInputStream() (or if that call succeeds) then the code is able to delete the file.  The delete only fails when 

UnzipUtil.createZipInputStream() fails to open the zip file.

Sent from Yahoo Mail for iPad


On Thursday, September 5, 2019, 1:36 AM, Srikanth Reddy Lingala <notifications@github.com> wrote:


I used the following code to test, and this code deletes the zip file. I tested it on a mac though. Which os are you using? Does your tests (or user running the tests) have the rights to delete a file?
ZipFile zipFile = new ZipFile("/Users/slingala/Downloads/test.zip", "password".toCharArray());
    ZipParameters zipParameters = new ZipParameters();
    zipParameters.setEncryptFiles(true);
    zipParameters.setEncryptionMethod(EncryptionMethod.AES);
    zipFile.addFile(TestUtils.getTestFileFromResources("sample.pdf"), zipParameters);

    try {
      zipFile = new ZipFile("/Users/slingala/Downloads/test.zip");
      zipFile.extractAll("/Users/slingala/Downloads/");
    } catch (Exception e) {
      assertThat(zipFile.getFile().delete()).isTrue();
      throw e;
    }

**Comment by srikanth-lingala** [comment]
I was able to reproduce on Windows with the below code:

```
@Test
  public void zipTest() throws IOException {

    ZipFile zipFile = new ZipFile("c:\\test.zip", "password".toCharArray());
    ZipParameters zipParameters = new ZipParameters();
    zipParameters.setEncryptFiles(true);
    zipParameters.setEncryptionMethod(EncryptionMethod.AES);
    zipFile.addFile(TestUtils.getTestFileFromResources("sample.pdf"), zipParameters);

    try {
      zipFile = new ZipFile("c:\\test.zip"); // intentionally not providing password to generate an exception
      try(InputStream inputStream = zipFile.getInputStream(zipFile.getFileHeaders().get(0))) {
        inputStream.read(new byte[4096]);
      }
    } catch (Exception e) {
      assertThat(zipFile.getFile().delete()).isTrue();
      throw e;
    }
  }
```

I will include a fix for it.

**Comment by srikanth-lingala** [comment]
Fixed in v2.1.3 released today
